### ISSUE TITLE
Crash in using trained inverse model in cpu mode.

### ISSUE THREAD

**Issue by logic1988** [rtqichen/style-swap#7]
When I using trained inverse model in cpu mode to generate stylized image by following command:
`th style-swap.lua --content images/content/bike.jpg --style images/style/starry_night.jpg --decoder vgginv/dec-045500.t7 -cpu`
There are the following crashes,  but in GPU mode, everything is normal.
```
/Users/logic/Desktop/torch/install/bin/luajit: ...gic/Desktop/torch/install/share/lua/5.1/nn/Container.lua:67:
In 2 module of nn.Sequential:
In 1 module of nn.Sequential:
...rs/logic/Desktop/torch/install/share/lua/5.1/nn/THNN.lua:109: bad argument #3 to 'v' (cannot convert 'struct THFloatTensor *' to 'struct THDoubleTensor *')
stack traceback:
	[C]: in function 'v'
	...rs/logic/Desktop/torch/install/share/lua/5.1/nn/THNN.lua:109: in function 'SpatialConvolutionMM_updateOutput'
	...op/torch/install/share/lua/5.1/nn/SpatialConvolution.lua:112: in function <...op/torch/install/share/lua/5.1/nn/SpatialConvolution.lua:107>
	[C]: in function 'xpcall'
	...gic/Desktop/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	...ic/Desktop/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function <...ic/Desktop/torch/install/share/lua/5.1/nn/Sequential.lua:41>
	[C]: in function 'xpcall'
	...gic/Desktop/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	...ic/Desktop/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
	style-swap.lua:203: in function 'swapTransfer'
	style-swap.lua:247: in main chunk
	[C]: in function 'dofile'
	...ktop/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x0104591cf0

WARNING: If you see a stack trace below, it doesn't point to the place where this error occured. Please use only the one above.
stack traceback:
	[C]: in function 'error'
	...gic/Desktop/torch/install/share/lua/5.1/nn/Container.lua:67: in function 'rethrowErrors'
	...ic/Desktop/torch/install/share/lua/5.1/nn/Sequential.lua:44: in function 'forward'
	style-swap.lua:203: in function 'swapTransfer'
	style-swap.lua:247: in main chunk
	[C]: in function 'dofile'
	...ktop/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:145: in main chunk
	[C]: at 0x0104591cf0
```

**Comment by rtqichen** [comment]
Thanks for letting me know. CPU mode was using both doubles and floats. This is now fixed in 738550406ee4ba12baa0b2814c2619a3063df4f4.

**Comment by logic1988** [comment]
@rtqichen Thank you very much. 

Training the model is an extremely time-consuming process, I used about 50000 content images and more than 200 style images to train, spent more than 4 days.  But I founded the result was not as good as original result, always the effect is very poor as following: 
![golden_gate_stylized](https://cloud.githubusercontent.com/assets/13864726/21833255/d978bf6a-d7eb-11e6-8bf0-e5ef1a8cc2e3.jpg)
![bike_stylized](https://cloud.githubusercontent.com/assets/13864726/21833257/df42eefc-d7eb-11e6-9682-f428070751aa.jpg)

Will the reason is because the number of style image is too small?
"Painter by Numbers" database was too large. How long did it take to train as described in the paper (both 80,000 images)? Can you provide a good model of your training?

**Comment by rtqichen** [comment]
The reason is likely because your inverse network was trained with maxpooling layers, but it's being evaluated with average pooling layers. Be sure to specify `--pooling max` when using your inverse net trained with the current script. Having less images should actually make it easier to train (or at least easier to overfit).

Yes, it takes a very long time to train. I haven't provided a trained model because I'm not quite satisfied with my current results either. I'm thinking of training an inverse network with more advanced architectures (resnet, subpixel conv layers) but haven't gotten around to it due to other deadlines..

Example using one of the better networks I trained:

Optimization
![golden_gate_optim](https://cloud.githubusercontent.com/assets/3375899/21834807/ff75faea-d76c-11e6-8fc1-de35fb50e06e.jpg)

Inverse net with transposed convolutions, instance normalization, and sigmoid squashing
![golden_gate_starry_invnet](https://cloud.githubusercontent.com/assets/3375899/21834808/02178f16-d76d-11e6-8e0b-d83de457955a.jpg)

Same network, but using average pooling at test time. 
![golden_gate_starry_invnet_wrongpooling](https://cloud.githubusercontent.com/assets/3375899/21834809/03f2b270-d76d-11e6-9458-66c05769513c.jpg)

This invnet tends to make everything darker due to the Painter by Numbers dataset having a bias towards darker textures. Starry Night seems to be one of those styles that just aren't easy to do with an invnet. The invnet assumes the activations it gets came from a neural network with max pooling layer, so when used with activations that came from average pooling layers, it fails at rendering a lot of textures.

**Comment by logic1988** [comment]
@rtqichen I tested my trained model by specify `--pooling max`, and the result was also not so good.
I trained by `train-vgg-decoder.lua` in default setting, It seems that `--instanceNorm` and `--tconv `  were not turned on by default.  And I did not find `sigmoid squashing` setting.

**Comment by xiesiyuan** [comment]
--instanceNorm is very important for quality.
